### ISSUE TITLE
GuzzleHttp error when using wp2static-addon-s3

### ISSUE THREAD

**Issue by tweety101** [leonstafford/lokl#13]
Hi, I really like your LOKL setup and script. 
This runs way faster than my own WordPress docker setup. 

When I try to use the wp2static-addon-s3, I keep getting a PHP error: 
"PHP message: PHP Fatal error:  Uncaught Error: Call to undefined function GuzzleHttp\_idn_uri_convert() in /usr/html/wp-content/plugins/wp2static-addon-s3/vendor/guzzlehttp/guzzle/src/Client.php:220"

This is in a fresh container built with the v4 version of lokl ( \sh -c "$(curl -sSl 'https://lokl.dev/go?v=4')")

I found [this](https://github.com/aws/aws-sdk-php/issues/2006) which indicates it could be caused by two different versions of Guzzle being installed

However, I can't figure out how to fix it. 

Thanks for any advice!

**Comment by leonstafford** [comment]
Hi @tweety101 - sorry for the slow response!

Thanks for the feedback! I'm loving this Lokl thing every chance I get to use it, too! 

Do you have other plugins installed in that instance where you get the issue? 

That link was promising. [This one](https://github.com/guzzle/guzzle/issues/2511) also has some more insights/ways to debug further. 

Did you do any other Composer installs that may have installed an older global version? 

PS. I've started some work on a PHP 8 version of Lokl, which should be easy enough to keep separate from PHP 7 with different tags. Nothing published yet though

**Comment by tweety101** [comment]
Hi @leonstafford,

I get the error when I do a fresh install using the script and then enable the S3 output in wp2static and run the generator. 
The container is built from the image: lokl/lokl:0.0.19

Are you not seeing the same thing? Maybe I have an issue with my docker images. 

PHP8 would be awesome by the way. Looking forward to that.

**Comment by leonstafford** [comment]
@tweety101 I haven't done a full test of S3 in a while, so that may be it. I did bump the AWS SDK version this week in the S3 addon, maybe that will sort this out. But it's not in Lokl yet, I've got a bit more work to do to get all addons PHP8 compatible and baked into Lokl (and offer PHP 7 and 8 versions of Lokl vs force everyone to 8).

WP2Static's core plugin now looks PHP8 compatible after some work last night. Coincidentally, that was replacing a lot of cURL commands in crawling with more Guzzle!

Short term workaround for you: use zip or no deployment option and install AWS CLI via Lokl's SSH (if that's in the V4 you're using, it's in the branch I've been working in) and run the `aws s3 sync` command to get the files in, along with a CloudFront invalidation. This is what I use for all my Hugo sites, like https://github.com/leonstafford/ljsdotdev/blob/8460bddb144bb62a5654b6fa48b2d2252b9eac1b/deploy.sh#L9-L11

One other thought, maybe I do have a global Guzzle in there somewhere, if by chance I used composer to install WP_CLI or such within Lokl, will need to check on that...

**Comment by leonstafford** [comment]
about to do some new tests now in the php 8 Lokl I'm working on with a fresh build of S3 addon and will see how it goes. Then wil backtrack lokl a bit and split into PHP7 and 8 tags and do full S3/CF deployment test

**Comment by leonstafford** [comment]
so far so good, with action/enabling of S3 fine, crawling looks good.

new issue to debug, my docker container cannot resolve `s3.amazonaws.com`...  host machine can (works on part of my machine!)

will get over that and then continue with the deploy testing

**Comment by leonstafford** [comment]
fun, looks like [this](https://forums.docker.com/t/intermittent-dns-resolving-issues/9584/2)

**Comment by leonstafford** [comment]
OK, almost there :) using `8.8.4.4` in `resolv.conf` within container feels a bit odd, but as our Lokl containers aren't trying to access anything on host or other containers, this may be a good permanent change...

dealing with small issue with s3 addon now, likely php8 compat issue:

```
2020/12/06 02:04:22 [error] 65#65: *68 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Error: Call to undefined function WP2StaticS3\finfo_open() in /usr/html/wp-content/plugins/wp2static-addon-s3/src/MimeTypes.php:1017
Stack trace:
#0 /usr/html/wp-content/plugins/wp2static-addon-s3/src/Deployer.php(88): WP2StaticS3\MimeTypes::guessMimeType()
#1 /usr/html/wp-content/plugins/wp2static-addon-s3/src/Controller.php(266): WP2StaticS3\Deployer->uploadFiles()
#2 /usr/html/wp-includes/class-wp-hook.php(287): WP2StaticS3\Controller->deploy()
#3 /usr/html/wp-includes/class-wp-hook.php(311): WP_Hook->apply_filters()
#4 /usr/html/wp-includes/plugin.php(484): WP_Hook->do_action()
#5 /usr/html/wp-content/plugins/wp2static/src/Controller.php(640): do_action()
#6 /usr/html/wp-content/plugins/wp2static/src/Controller.php(730): WP2Static\Controller::wp2staticHeadless()
#7 /usr/html/wp-includes/class-wp-hook.php(285): WP2Static\Controller::wp2staticRun()
#8 /usr/html/wp-includes/class-wp-hook.php(311): WP_Hook->apply_filters()
#9 /usr/html/wp-includes/plugin.php(484): W" while reading response header from upstream, client: 172.17.0.1, server: , request: "POST /wp-admin/admin-ajax.php HTTP/1.1", upstream: "fastcgi://unix:/var/run/php8-fpm.sock:", host: "localhost:4437", referrer: "http://localhost:4437/wp-admin/admin.php?page=wp2static"
```

Sorry for rambling here, just talking to myself :D

**Comment by tweety101** [comment]
Hahaha, no worries:D

Interesting that you don't see the error. I will try deleting the docker images so a new one is pulled. 

For now, I am using the older version of the plugin: Static HTML

**Comment by leonstafford** [comment]
@tweety101 - an update, I'm taking this approach: https://github.com/leonstafford/AriseByAnyOtherName to isolate our copy of Guzzle from any conflicts with other versions. I'll also be doing the same for Static HTML Output, so they'll each use a properly isolated version. The only major 3rd party libraries we're using tend to be that, the AWS SDK, and possibly a few of the deployers which provide a PHP SDK. 

Will update this once cleared

**Comment by leonstafford** [comment]
This should be closed now in latest release (5.0.0-rc2), which is now on homepage's instructions. Please let me know if any issues deploying to S3/CloudFront appear.
